### PR TITLE
[2/N][trainer] feat: bagel flowgrpo integration 

### PR DESCRIPTION
**Pull request by zhtmike**
### What does this PR do?

follow up of #66; validated on vllm-omni==0.20.0 only

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

螢幕截圖 2026-06-12 下午2 52 56
螢幕截圖 2026-06-12 下午2 53 16
螢幕截圖 2026-06-12 下午2 54 40

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `diffusion`, `omni`, `tests`, `docker`
  - If this PR involves multiple modules, separate them with `,` like `[diffusion, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][diffusion, fsdp] feat: new rollout scheduler`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update the documentation.
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Review by SamitHuang — COMMENTED** [review]
Please update the README.md as well

File: verl_omni/pipelines/bagel_flow_grpo/vllm_omni_rollout_adapter.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        traj_latents = output.trajectory_latents
+        traj_timesteps = output.trajectory_timesteps
+        traj_log_probs = output.trajectory_log_probs
+
+        if sde_window is not None:
+            begin, end = sde_window
+            if traj_latents is not None:
+                traj_latents = traj_latents[begin : end + 1]
+            if traj_timesteps is not None:
+                traj_timesteps = traj_timesteps[begin:end]
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl_omni/pipelines/bagel_flow_grpo/vllm_omni_rollout_adapter.py:327]: not relavent

File: verl_omni/pipelines/bagel_flow_grpo/vllm_omni_rollout_adapter.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+        if isinstance(extra_args.get("cfg_interval"), list):
+            extra_args["cfg_interval"] = tuple(extra_args["cfg_interval"])
+
+        # Pick SDE window: noise and log-prob recording only inside this range.
+        logprobs = bool(extra_args.get("logprobs", True))
+        noise_level = float(extra_args.get("noise_level", 0.0))
+        sde_window_size = extra_args.get("sde_window_size", None)
+        sde_window_range = extra_args.get("sde_window_range", None)
+        if isinstance(sde_window_range, list):
+            sde_window_range = tuple(sde_window_range)
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl_omni/pipelines/bagel_flow_grpo/vllm_omni_rollout_adapter.py:288]: not relavent

File: examples/flowgrpo_trainer/bagel_deploy_config.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# BAGEL-7B-MoT single-stage deploy config for FlowGRPO training.
```
Comment (AndyZhou952) [inline examples/flowgrpo_trainer/bagel_deploy_config.yaml:1]: Later, we may consider separate subfolders by models

File: verl_omni/pipelines/bagel_flow_grpo/vllm_omni_rollout_adapter.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+
+
+def _to_token_list(token_ids: Any) -> list[int] | None:
+    if token_ids is None:
+        return None
+    if isinstance(token_ids, torch.Tensor):
+        token_ids = token_ids.detach().cpu().tolist()
+    if token_ids and isinstance(token_ids[0], list):
+        token_ids = token_ids[0]
+    return [int(token_id) for token_id in token_ids]
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply verl_omni/pipelines/bagel_flow_grpo/vllm_omni_rollout_adapter.py:63]: bagel only. works fine.

File: examples/flowgrpo_trainer/data_process/bagel_ocr.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+import datasets
+import numpy as np
+from transformers import AutoTokenizer
+from verl.utils.hdfs_io import copy, makedirs
+
+
+def extract_ocr_solution(text: str) -> str:
+    """Extract the ground-truth text from OCR solution string."""
+    return text.split('"')[1]
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/flowgrpo_trainer/data_process/bagel_ocr.py:45]: not relavant.

File: verl_omni/pipelines/bagel_flow_grpo/bagel_model.py [hunk 6]
```diff
@@ -0,0 +1,644 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""BagelForTraining – FSDP-compatible BAGEL MoT module for flow-matching training.
+
+Ported from vllm-omni/BAGEL.  MoT dual pathways (text vs generation),
```
Comment (AndyZhou952) [inline verl_omni/pipelines/bagel_flow_grpo/bagel_model.py:17]: A little bit curious why we define it here but not upstream

**Comment by SamitHuang** [comment]
@princepride PTAL

File: examples/flowgrpo_trainer/data_process/bagel_ocr.py
Comment (SamitHuang) [inline examples/flowgrpo_trainer/data_process/bagel_ocr.py:1]: Can add a doc for running bagel?

**Review by AndyZhou952 — APPROVED** [review]
Overall LGTM

**Comment by princepride** [comment]
LGTM

File: examples/flowgrpo_trainer/data_process/bagel_ocr.py
Comment (zhtmike) [reply examples/flowgrpo_trainer/data_process/bagel_ocr.py:1]: next PR. together with vllm-omni 0.22 verification; and doc of how to integrate non-diffusers model

File: examples/flowgrpo_trainer/bagel_deploy_config.yaml (hunk 3 above)
Comment (zhtmike) [reply examples/flowgrpo_trainer/bagel_deploy_config.yaml:1]: agree

File: verl_omni/pipelines/bagel_flow_grpo/bagel_model.py (hunk 6 above)
Comment (zhtmike) [reply verl_omni/pipelines/bagel_flow_grpo/bagel_model.py:17]: because there is no third-party lib supports bagel now. possibly in future veomini will support this

**Review by SamitHuang — APPROVED** [review]
LGTM
